lib/rust/flexer/src/lib.rs

```diff
@@ -14,26 +15,220 @@
 //! These lexers are capable of lexing any regular grammar, with some extensions to allow working
 //! with context sensitive (e.g. indentation-aware) syntax.
 
-pub use enso_prelude as prelude;
+// TODO [AA] Logging https://github.com/enso-org/ide/blob/main/src/rust/ide/src/model/execution_context/synchronized.rs#L45
 
+use crate::prelude::*;
 
+use lazy_reader::LazyReader;
+use crate::group::Group;
 
 pub mod automata;
-pub mod group;
 pub mod data;
+pub mod group;
 
+#[allow(missing_docs)]
+pub mod prelude {
+    pub use enso_prelude::*;
+}
 
 
-// ===============
-// === Flexer ====
-// ===============
+
+// ==============
+// === Flexer ===
+// ==============
 
 /// The flexer is an engine for generating lexers.
 ///
 /// Akin to flex and other lexer generators, it is given a definition as a series of rules from
 /// which it then generates code for a highly optimised lexer implemented on top of a
 /// [DFA](https://en.wikipedia.org/wiki/Deterministic_finite_automaton).
-pub trait Flexer {
+#[derive(Clone,Debug)]
+pub struct Flexer<Definition,Output,Reader> where Reader:LazyReader {
+    /// The stack of states that are active during lexer execution.
+    pub state_stack: Vec<usize>,
+    /// A reader for the input.
+    pub reader: Reader,
+    /// The current match of the lexer.
+    pub current_match: String,
```

Comment:
how does it work? I dont see any code that modifies `current_match` field in this PR.

Reply:
It's actually used purely by the user impl, so I've moved it there.

Reply:
Having taken another look at this, the presence of it is something that'll be relied on by the code generator, so either it needs to become part of the state interface, or it should live in the Flexer.

Reply:
Note to self: move this back.